### ISSUE TITLE
Query Failure

### ISSUE THREAD

**Issue by kimdotnet** [IBM-Blockchain-Archive/marbles#17]
I deployed marbles in localhost, everything run fine until some of time.
When I wake up this morning, and tried to run marbles again, I got this error:

```
read  - success: { jsonrpc: '2.0',
  error:
   { code: -32003,
     message: 'Query failure',
     data: 'Error when querying chaincode: Error:Failed to launch chaincode spec(Timeout expired while starting chaincode 14b711be6f0d00
b190ea26ca48c22234d93996b6e625a4b108a7bbbde064edf0179527f30df238d61b66246fe1908005caa5204dd73488269c8999276719ca8b(networkid:dev,peerid:
c9527fe1-2332-48aa-8e6f-797d0689c52e_vp1,tx:f77a672d-b239-498b-9e54-803ebeccfca5))' },
  id: 1466646199791 }
marbles index msg error: [SyntaxError: Unexpected token u]
```

All marbles did I create also dissapear and I can't create new marble.
Any idea what happened?

**Comment by mastersingh24** [comment]
@kimdotnet   - are you running everything on localhost - including the peers?

**Comment by kimdotnet** [comment]
no, I use blockchain service for the peers.

Now I also got this error when trying to get marbles data, any idea what happen:
`[ibc-js] read  - success: { jsonrpc: '2.0',
  result: { status: 'OK', message: 'null' },
  id: 1466764314732 }`

**Comment by mastersingh24** [comment]
Unfortunately, I think you may just need to delete your instance of the blockchain service on BMX and create a new one.  We had a few issues which we have since corrected and unfortunately a few networks are not in an easily recoverable state.

**Comment by kimdotnet** [comment]
@mastersingh24 I already delete my instance, now that problem is solved. But now I got new problem.
My marbles is dissapear after some time (most likely 8 hours).

In my console I got this error, after deploy success message.

```
        deploy success [waiting another 50 seconds]
         14b711be6f0d00b190ea26ca48c22234d93996b6e625a4b108a7bbbde064edf0179527f30df238d61b66246fe1908005caa5204dd73488269c8999276719ca8b

[ibc-js] ibc.save() error { [Error: ENOENT: no such file or directory, open 'C:\Users\kim\Documents\Learn node.js\marbles\cc_summaries\14b711be6f0d00b190ea26ca48c22234d93996b6e625a4b108a7bbbde064edf0179527f30df238d61b66246fe1908005caa5204dd73488269c8999276719ca8b.json']
  errno: -4058,
  code: 'ENOENT',
  syscall: 'open',
  path: 'C:\\Users\\kim\\Documents\\Learn node.js\\marbles\\cc_summaries\\14b711be6f0d00b190ea26ca48c22234d93996b6e625a4b108a7bbbde064edf0179527f30df238d61b66246fe1908005caa5204dd73488269c8999276719ca8b.json' }
```

What the meaning of that error? is that error related to the loss of my marble?

**Comment by dshuffma-ibm** [comment]
That error is harmless and would have nothing to do with your symptoms.  It simply cannot save an optional summary file that can be used to debug/transfer ibc settings.

As for losing your marbles, do you notice if the blockchain ledger is resetting? Examine the block height of the chain from either the UI at the bottom, or by using the GET /chain API directly.  Is it possible someone else is using your network?

**Comment by kimdotnet** [comment]
@dshuffma-ibm okay, after trial and error. Now I got this pattern.
1. Run gulp command
2. insert marble
3. exit program (by pressing ctrl+c)
4. run gulp again
5. marbles are still there

This is the case:
My marbles only lost when I run "gulp" command while the CC status is "Exited (0) X hours ago". Any idea why this is happen?
My marble is safe whenever I run gulp comman, but CC status is "UP".

**Comment by dshuffma-ibm** [comment]
interesting, so you are saying:
- everything is okay if you run gulp when the status for the chaincode container is "up" 
- previous marbles are lost if you run gulp when the status is exited

I will try this out myself in a few minutes and let you know.

**Comment by dshuffma-ibm** [comment]
Well there is something going on. I was able to create 1 network to mimic exactly what you saw. Unfortunately I can't seem to do it again!

For some unknown reason its possible for the chaincode container to be stuck in a created state.  Its not running, but its existence means the peer will be unable to bring up another. (to clarify this is not supposed to happen, and it seems rare.  chaincode is suppose to exit after inactivity and then restart when needed)

A temporary fix is we can delete the stuck container if you provide your network ID.  The long term fix is we have noted this bug and will attempt to make it repeatable so we can fix it.

**Comment by kimdotnet** [comment]
@dshuffma-ibm 

> interesting, so you are saying:

```
everything is okay if you run gulp when the status for the chaincode container is "up"
previous marbles are lost if you run gulp when the status is exited
```

yes, exactly. Btw, thank you for your answer.

# 

One more question david, Is it possible to send JSON error message from chaincode? if it is possible, is it a good practice or not?

For example:
In case I adding marble, which the marble already exist in the chaincode state. Is it possible to send this json:

```
{
  "jsonrpc": "2.0",
  "result": {
    "status": "FAIL",
    "message": "Marble already exists"
  },
  "id": 0
}
```

rather than this:

```
{
  "jsonrpc": "2.0",
  "result": {
    "status": "OK",
    "message": "c124a17a-7a6a-406c-bf9f-125a8cf8326b"
  },
  "id": 0
}
```

Because in marble app, if I tried to add/init 2 or more marble with the same name. There is no error message for client to know, like "Hey, this marble already exists".
In current state, marble app only will redirect to home page. Regardless of whether marble successfully added to the chain or not.

I am sure, it will be better if client know whether his action is success or not.

**Comment by dshuffma-ibm** [comment]
So we need to separate the question into two parts.

_Is it possible to return an error from a query cc function?_
- Yes!. The chaincode should return an error such as:

```
return nil, errors.New("Something is wrong. Error details and such here as a string")
```
- this should be visible in the response from a query API call, I believe its buried in the `response.error.message` JSON (might not be exact field name, but its close)
- I would consider this excellent practice.  Getting an error msg back directly is a lot faster than burying your self in log messages to see what happened.
- see [marbles chaincode](https://github.com/IBM-Blockchain/marbles-chaincode/blob/master/hyperledger/part1/part1_chaincode.go) for examples (line 143 is one)

---

_Is it possible to return an error from an invoke cc function?_
- No. The outcome of an invoke is not decided at the time of the API call. In the _real world_ there is a consensus algorithm to go through. Therefore the success/failure of the chaincode is not known until its executed and there is consensus on the transaction.
- tldr: You will always see _invoke successful_
- there has been talks about adding callbacks via event subscriptions to determine the status of a transaction, but as far as I know it is not yet implemented

**Comment by dionalfa** [comment]
@dshuffma-ibm okay, I get it. So at this time, it's not possible to return error message from invocation.
Hope it will be implemented later, because it will be confusing for the user if he can't know whether his action is successfully saved or not. example: when adding marble which is already exists.

Thank you for your support david.

**Comment by kimdotnet** [comment]
@dshuffma-ibm  okay I get it. So at this time, it's not possible to return an error message from invocation.
Hope it will be implemented later, because it will be confusing for the user if he can't know whether his action is successfully saved or not. Example: when adding marble which is already exists.

Thank you for your support david.
